packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +465,106 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    let hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => host.$call('assert_can_evacuate')))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => {
+          const hostId = vm.$resident_on?.$id
+
+          if (hostId === undefined) {
+            throw new Error('Could not find host of all running VMs')
+          }
+
+          return hostId
+        }
+      ),
+      vms => vms.map(vm => vm.$id)
+    )
+
+    // Cache $fields to make sure we can access them later
+    hosts = hosts.map(host => ({
+      ...host,
+      id: host.$id,
+      ref: host.$ref,
+      metricsId: host.$metrics.$id,
+      metricsRef: host.$metrics.$ref,
+    }))
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.ref === this.pool.master)
+    if (indexOfMaster > 0) {
+      ;[hosts[0], hosts[indexOfMaster]] = [hosts[indexOfMaster], hosts[0]]
+    }
+
+    // Restart all the hosts one by one
+    for (const host of hosts) {
+      const hostId = host.id
+      const metricsId = host.metricsId
+
+      await this.barrier(host.metricsRef)
+      await this._waitObjectState(metricsId, metrics => metrics.live)
+      const rebootTime = parseDateTime(await this.call('host.get_servertime', host.ref))
+
+      log.debug(`Evacuate and restart host ${hostId}`)
+      await this.rebootHost(hostId)
+
+      log.debug(`Wait for host ${hostId} to be up`)
+      await timeout.call(
+        (async () => {
+          await this._waitObjectState(
+            hostId,
+            host => host.enabled && rebootTime < host.other_config.agent_start_time * 1e3
+          )
+          await this._waitObjectState(metricsId, metrics => metrics.live)
+        })(),
+        5 * 60 * 1e3,
```

Comment:
Please move this timeout to the `xo-server` configuration and humanize it using `parseDuration`.